bin/network-monitor/src/counter.rs

```diff
@@ -0,0 +1,429 @@
+    NoteScript,
+    NoteTag,
+    NoteType,
+};
+use miden_objects::transaction::{InputNotes, PartialBlockchain, TransactionArgs};
+use miden_objects::utils::Deserializable;
+use miden_objects::{Felt, Word, ZERO};
+use miden_tx::auth::BasicAuthenticator;
+use miden_tx::utils::Serializable;
+use miden_tx::{LocalTransactionProver, TransactionExecutor};
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use tokio::sync::watch;
+use tokio::time::sleep;
+use tracing::{error, info, instrument};
+
+use crate::COMPONENT;
+use crate::config::MonitorConfig;
+use crate::deploy::account::{MonitorDataStore, get_counter_library};
+use crate::deploy::wallet::create_wallet_account;
+use crate::status::{ServiceDetails, ServiceStatus, Status};
+
+/// Counter increment task details.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Default)]
+pub struct CounterIncrementDetails {
+    /// Number of successful counter increments.
+    pub success_count: u64,
+    /// Number of failed counter increments.
+    pub failure_count: u64,
+    /// Current counter value (if available).
+    pub current_value: Option<u64>,
+    /// Last transaction ID (if available).
+    pub last_tx_id: Option<String>,
+}
+
+async fn create_rpc_client(config: &MonitorConfig) -> Result<RpcClient> {
+    Builder::new(config.rpc_url.clone())
+        .with_tls()
+        .context("Failed to configure TLS for RPC client")
+        .unwrap()
+        .with_timeout(Duration::from_secs(30))
+        .without_metadata_version()
+        .without_metadata_genesis()
+        .connect::<Rpc>()
+        .await
+}
+
+async fn get_genesis_block_header(rpc_client: &mut RpcClient) -> Result<BlockHeader> {
+    // Get the latest block header
+    let block_header_request = BlockHeaderByNumberRequest {
+        block_num: Some(BlockNumber::GENESIS.as_u32()), // Get latest block
+        include_mmr_proof: None,
+    };
+
+    let response = rpc_client
+        .get_block_header_by_number(block_header_request)
+        .await
+        .context("Failed to get latest block header from RPC")?
+        .into_inner();
+
+    let latest_block_header = response
+        .block_header
+        .ok_or_else(|| anyhow::anyhow!("No block header in response"))?;
+
+    let block_header: BlockHeader =
+        latest_block_header.try_into().context("Failed to convert block header")?;
+
+    Ok(block_header)
+}
+
+/// Wait until at least one block after `after` is available.
+async fn wait_for_block_after(rpc_client: &mut RpcClient, after: BlockNumber) -> Result<()> {
+    let target = after.as_u32().saturating_add(1);
+    loop {
+        let req = BlockHeaderByNumberRequest {
+            block_num: Some(target),
+            include_mmr_proof: None,
+        };
+        let resp = rpc_client.get_block_header_by_number(req).await?.into_inner();
+        if resp.block_header.is_some() {
+            return Ok(());
+        }
+        sleep(Duration::from_millis(500)).await;
+    }
```

Comment:
Do we want a timeout for this loop? Perhaps even just a log if the timeout is reached (if we don't want to error out).